torch/csrc/jit/tensorexpr/kernel.cpp

```diff
@@ -455,6 +509,10 @@ void TensorExprKernel::promoteInputs(std::vector<ExprHandle>& inputs) {
     highType = promoteTypes(highType, input.dtype().scalar_type());
   }
 
+  if (!checkTypes(highType, typeConstraints)) {
+    throw unsupported_dtype();
```

Comment:
IMO we should stop throwing exceptions for unsupported cases since the fallback is disabled; we either need to handle things properly at this point, or clean things out during tensorexpr_fuser.

Reply:
4 places in the original `kernel.cpp` throws `unsupported_dtype()` (i.e., `promoteDType`, `demoteDType`, `constant`, `bindInput`. There are two reasons for unsupported dtypes:
- **User errors**, i.e., users specify invalid data types for a kernel function (e.g., `torch.frac()` does not take floating-point as input
- **Out-of-sync data type support** compared to eager mode, i.e., data type support in the JIT will lag behind that of eager runtime from time to time. I suspect this may be the reality, for instance, `kernel.cpp` does not seem to be no support for complex and qint types, and support for half floating-point may be lacking too (in `test_jit_fuser_te.py`, `test_unsupported_dtypes()` seems to give way to which types were considered as unsupported right now). This may be the direct consequence of how PyTorch rollout supports for new kernels or new types (eager mode first, then torch script, JIT, other devices, etc).  If we expect PyTorch kernels to continue evolving, there may always be a catch-up game.

If unsupported data types do occur in reality, I (as a developer) much prefer that the detection is explicit. As a user though, I'd prefer the error messages could differentiate between my (user) errors and the JIT's internal errors. Right now, `checkTypes` does not differentiate these two types of errors. For instance, for `frac` it will report unsupported data types for integer types (which will be user errors), but it will also report unsupported types for complex (actually I am not sure whether it is user errors or internal errors).

Reply:
Regarding un-even data type support, even the eager mode does not seem to support all data types in one batch. For instance, `torch.f16` is supported for `frac()` but not `sin()` (unless `sin()` is designed not to take `f16`?). If that is the case, then the data type support in PyTorch would be even more fluid.

Reply:
Right so you're absolutely correct that there are currently places in `kernel.cpp` that throw `unsupported_dtype` but I think they should go away and we should either (1) actually support that dtype, or (2) filter those cases out in the `tensorexpr_fuser` pass.  As it stands right now, we'll attempt to fuse things that aren't supported, fail, and then the user is completely blocked, even though eager mode could handle the case.